shell/client.go

```diff
@@ -0,0 +1,329 @@
+	return cmd
+}
+
+var stdReadWriter = struct {
+	io.Reader
+	io.Writer
+}{
+	Reader: os.Stdin,
+	Writer: os.Stdout,
+}
+
+func dialShell(c Config, sigs <-chan os.Signal, w io.Writer) (net.Conn, error) {
+	var conn net.Conn
+	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
+	defer cancel()
+	for {
+		var err error
+		var d net.Dialer
+		conn, err = d.DialContext(ctx, "unix", c.ShellSockPath)
+		if err == nil {
+			break
+		}
+		// Dialing failed. Server might not be fully up yet. Wait a bit and retry.
+		select {
+		case <-sigs:
+			return nil, fmt.Errorf("interrupted")
+		case <-ctx.Done():
+			return nil, fmt.Errorf("dialing timed out: %w", err)
+		case <-time.After(time.Second):
+			fmt.Fprintf(w, "Dialing failed: %s. Retrying...\n", err)
+		}
+	}
+	return conn, nil
+}
+
+// ShellExchange sends a single command to the shell. Output is written
+// to the given writer [w].
+func ShellExchange(c Config, w io.Writer, format string, args ...any) error {
+	conn, err := dialShell(c, nil, os.Stderr)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+	_, err = fmt.Fprintf(conn, format+"\nexit\n", args...)
+	if err != nil {
+		return err
+	}
+	bio := bufio.NewReader(conn)
+	for {
+		lineBytes, isPrefix, err := bio.ReadLine()
+		if err != nil {
+			return nil
+		}
+		line := string(lineBytes)
+		if line == stdoutMarker || line == stderrMarker {
+			// Commands that write to "stdout" instead of the log show the [stdout] as
+			// the first line. This is useful information in tests, but not useful in
+			// the shell, so just skip this.
+			continue
+		}
+		line, ended := strings.CutSuffix(line, endMarker)
+		if isPrefix {
+			// Partial line, don't print \n yet.
+			_, err = fmt.Fprint(w, line)
+		} else {
+			_, err = fmt.Fprintln(w, line)
+		}
+		if err != nil {
+			return err
+		}
+		if ended {
+			return nil
+		}
+	}
+}
+
+func executeShell(cfg Config, prompt string, printGreeting func(io.Writer), args []string) {
+	if len(args) > 0 {
+		err := ShellExchange(cfg, os.Stdout, "%s", strings.Join(args, " "))
+		if err != nil {
+			fmt.Fprintf(os.Stdout, "error: %s\n", err)
+		}
+	} else {
+		os.Exit(interactiveShell(cfg, prompt, printGreeting))
+	}
+}
+
+func interactiveShell(cfg Config, prompt string, printGreeting func(w io.Writer)) int {
+	// Try to set the terminal to raw mode (so that cursor keys work etc.)
+	restore, err := script.MakeRaw(0)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error setting terminal to raw mode: %s\n", err)
+	} else {
+		defer restore()
+	}
+
+	console := term.NewTerminal(stdReadWriter, prompt)
+	if width, height, err := term.GetSize(0); err == nil {
+		console.SetSize(width, height)
+	}
+	if printGreeting != nil {
+		printGreeting(console)
+	}
+
+	// Listen for SIGINT to stop.
+	sigs := make(chan os.Signal, 1)
+	defer func() {
+		signal.Stop(sigs)
+		close(sigs)
+	}()
+	signal.Notify(sigs, os.Interrupt)
+
+	var redial atomic.Bool
+	redial.Store(true)
```

Comment:
Maybe this is intentional, but the thing I tried to fix in https://github.com/cilium/cilium/pull/40950 was that typing `exit` in a working shell would immediately reconnect. Given `redial` doesn't seem to be set to false except after redialing, I'd expect this to be a problem again? Didn't have time to set up a test for it, though.